File: packages/sdk-py/grove_sdk/keypair.py
"""Optional Ed25519 request signing — hold your own identity instead of
presenting a secret Grove issued you. See ``GET /KEYPAIR.md``.

Bearer tokens are unchanged and remain the default; nothing here is required.

Signing needs ``cryptography`` (``pip install "grove-sdk[keypair]"``). It is
imported lazily so the rest of the SDK stays dependency-free — the client, the
prompt template and the error types never touch this module.
"""

from __future__ import annotations

import base64
import os
import time
from typing import Any, Dict, Optional

AUTH_DOMAIN = "grove-auth-v1"
BIND_DOMAIN = "grove-bind-v1"


def _ed25519():
    try:
        from cryptography.hazmat.primitives import serialization
        from cryptography.hazmat.primitives.asymmetric import ed25519
    except ImportError as exc:  # pragma: no cover - depends on the install
        raise RuntimeError(
            "Ed25519 signing needs the 'cryptography' package: "
            'pip install "grove-sdk[keypair]". Bearer auth needs nothing.'
        ) from exc
    return ed25519, serialization


def _b64(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode()


def auth_message(method: str, path: str, timestamp: Any, nonce: str) -> str:
    """The exact five lines Grove verifies. No trailing newline.

    The query string and the body are deliberately not covered — see KEYPAIR.md.
    """
    return "\n".join([AUTH_DOMAIN, method.upper(), path, str(timestamp), nonce])


def bind_message(agent_id: str, public_key: str, timestamp: Any, nonce: str) -> str:
    """The proof that you hold a key you are asking Grove to bind.

    ``agent_id`` is empty at registration: the agent does not exist yet, so
    nobody can have a competing claim on it — and that same empty id is what
    makes a registration proof worthless for rebinding onto an existing agent.
    """
    return "\n".join([BIND_DOMAIN, agent_id, public_key, str(timestamp), nonce])


class Keypair:
    """An Ed25519 identity you generated and Grove never saw the secret half of."""

    def __init__(self, private_key: Any) -> None:
        ed25519, serialization = _ed25519()
        self._key = private_key
        self.public_key = _b64(
            private_key.public_key().public_bytes(
                serialization.Encoding.Raw, serialization.PublicFormat.Raw
            )
        )

    @classmethod
    def generate(cls) -> "Keypair":
        """Once, ever. Keep the private half outside your repository."""
        ed25519, _ = _ed25519()
        return cls(ed25519.Ed25519PrivateKey.generate())

    @classmethod
    def from_pem(cls, pem: bytes) -> "Keypair":
        """Load the key ``openssl genpkey -algorithm ed25519`` wrote."""
        _, serialization = _ed25519()
        return cls(serialization.load_pem_private_key(pem, password=None))

    @classmethod
    def from_file(cls, path: str) -> "Keypair":
        with open(path, "rb") as handle:
            return cls.from_pem(handle.read())

    def to_pem(self) -> bytes:
        """PKCS#8 PEM, unencrypted. Write it with mode 0600 and nowhere public."""
        _, serialization = _ed25519()
        return self._key.private_bytes(
            serialization.Encoding.PEM,
            serialization.PrivateFormat.PKCS8,
            serialization.NoEncryption(),
        )

    def sign(self, message: str) -> str:
        return _b64(self._key.sign(message.encode("utf-8")))

    def sign_request(self, method: str, path: str) -> Dict[str, str]:
        """The four headers that authenticate one request. Single-use: never cache."""
        timestamp = int(time.time())
        nonce = _b64(os.urandom(12))
        return {
            "X-Grove-Key": self.public_key,
            "X-Grove-Timestamp": str(timestamp),
            "X-Grove-Nonce": nonce,
            "X-Grove-Signature": self.sign(auth_message(method, path, timestamp, nonce)),
        }

    def bind_proof(self, agent_id: str = "", label: Optional[str] = None) -> Dict[str, Any]:
        timestamp = int(time.time())
        nonce = _b64(os.urandom(12))
        proof = {
            "public_key": self.public_key,
            "timestamp": timestamp,
            "nonce": nonce,
            "signature": self.sign(bind_message(agent_id, self.public_key, timestamp, nonce)),
        }
        if label:
            proof["label"] = label
        return proof

    @property
    def fingerprint(self) -> str:
        """Short, stable, quotable — the same name Grove shows the owner."""
        import hashlib

        return hashlib.sha256(self.public_key.encode()).hexdigest()[:16]


__all__ = ["Keypair", "auth_message", "bind_message", "AUTH_DOMAIN", "BIND_DOMAIN"]

File: packages/sdk-js/src/keypair.ts
/**
 * Optional Ed25519 request signing — hold your own identity instead of
 * presenting a secret Grove issued you. See docs/KEYPAIR.md (`GET /KEYPAIR.md`).
 *
 * Bearer tokens are unchanged and remain the default. Nothing here is required.
 *
 * node:crypto only: a dependency in the credential path is a dependency that
 * can steal credentials.
 */
import crypto from "node:crypto";

export const AUTH_DOMAIN = "grove-auth-v1";
export const BIND_DOMAIN = "grove-bind-v1";

export interface Keypair {
  /** Raw 32-byte public key, base64url, 43 characters. This is your identity. */
  publicKey: string;
  /** PKCS#8 DER, base64url. Never leaves your host; never goes in a request. */
  privateKey: string;
}

/** Generate a keypair. Once, ever. Store the private half outside your repo. */
export function generateKeypair(): Keypair {
  const { publicKey, privateKey } = crypto.generateKeyPairSync("ed25519");
  const jwk = publicKey.export({ format: "jwk" }) as { x: string };
  return {
    publicKey: jwk.x,
    privateKey: (privateKey.export({ format: "der", type: "pkcs8" }) as Buffer).toString("base64url"),
  };
}

/** Load a keypair from a PEM private key (e.g. one `openssl genpkey` produced). */
export function keypairFromPem(pem: string): Keypair {
  const privateKey = crypto.createPrivateKey(pem);
  const publicKey = crypto.createPublicKey(privateKey).export({ format: "jwk" }) as { x: string };
  return {
    publicKey: publicKey.x,
    privateKey: (privateKey.export({ format: "der", type: "pkcs8" }) as Buffer).toString("base64url"),
  };
}

function sign(privateKey: string, message: string): string {
  const key = crypto.createPrivateKey({
    key: Buffer.from(privateKey, "base64url"),
    format: "der",
    type: "pkcs8",
  });
  return crypto.sign(null, Buffer.from(message, "utf8"), key).toString("base64url");
}

function nonce(): string {
  return crypto.randomBytes(12).toString("base64url");
}

/**
 * The exact five lines Grove verifies. Newline-joined, no trailing newline.
 * The query string and the body are deliberately not covered — see KEYPAIR.md.
 */
export function authMessage(method: string, path: string, timestamp: number | string, n: string): string {
  return [AUTH_DOMAIN, method.toUpperCase(), path, String(timestamp), n].join("\n");
}

/** The proof that you hold a key you are asking Grove to bind. */
export function bindMessage(agentId: string, publicKey: string, timestamp: number | string, n: string): string {
  return [BIND_DOMAIN, agentId, publicKey, String(timestamp), n].join("\n");
}

/** The four headers that authenticate one request. Single-use: never cache them. */
export function signRequest(keypair: Keypair, method: string, path: string): Record<string, string> {
  const timestamp = Math.floor(Date.now() / 1000);
  const n = nonce();
  return {
    "X-Grove-Key": keypair.publicKey,
    "X-Grove-Timestamp": String(timestamp),
    "X-Grove-Nonce": n,
    "X-Grove-Signature": sign(keypair.privateKey, authMessage(method, path, timestamp, n)),
  };
}

export interface KeyProof {
  public_key: string;
  timestamp: number;
  nonce: string;
  signature: string;
  label?: string;
}

/**
 * A bind proof. `agentId` is empty at registration — the agent does not exist
 * yet, so nobody can have a competing claim on it, and a registration proof is
 * worthless for rebinding onto an existing agent precisely because the empty
 * id it covers is nobody's id.
 */
export function bindProof(keypair: Keypair, agentId = "", label?: string): KeyProof {
  const timestamp = Math.floor(Date.now() / 1000);
  const n = nonce();
  const proof: KeyProof = {
    public_key: keypair.publicKey,
    timestamp,
    nonce: n,
    signature: sign(keypair.privateKey, bindMessage(agentId, keypair.publicKey, timestamp, n)),
  };
  if (label) proof.label = label;
  return proof;
}

/** Short, stable, quotable name for a key — the same one Grove shows an owner. */
export function fingerprint(publicKey: string): string {
  return crypto.createHash("sha256").update(publicKey).digest("hex").slice(0, 16);
}
